### PR TITLE
Respect the max concurrent requests limit

### PR DESCRIPTION
**Pull request by lidizheng**
Last week, I broke HEAD in https://github.com/grpc/grpc/pull/26588.

The root cause is that I wasn't aware of a `max_concurrent_request` limit (default is 1024), see [definition](https://github.com/grpc/proposal/blob/daa0c007a31aa2807b33dab3c07ccba131856918/A37-xds-aggregate-and-logical-dns-clusters.md#xds_cluster_impl_experimental-lb-policy).

https://github.com/grpc/grpc/blob/095045b31cf5fbccbd65e71d24617c50fce02d3c/src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_impl.cc#L294

This PR adds a throttle to the existing wait group design. It won't start new RPCs until there are less than `max_concurrent_request` of in-flight RPCs.

This bug in the testing framework wasn't triggered before, because the previous total RPC is around 1600, which means only ~800 call will be injected some delay, and it's within the default 1024 limit. After I bumped up the total RPC to ~2500, there will be 1250 delayed calls, hence some calls are getting dropped.

### CODE REVIEW

File: test/cpp/end2end/xds_end2end_test.cc [hunk 1]
```diff
@@ -2674,27 +2675,40 @@ class XdsEnd2endTest : public ::testing::TestWithParam<TestType> {
     // Variables for synchronization
     absl::Mutex mu;
     absl::CondVar cv;
+    size_t started = 0;
     size_t completed = 0;
     // Set-off callback RPCs
     for (size_t i = 0; i < num_rpcs; i++) {
+      {
+        // xDS LB policy `xds_cluster_impl_experimental` has a default
```
Comment (markdroth) [inline test/cpp/end2end/xds_end2end_test.cc:2683]: I don't think this makes sense.  The whole point of `SendConcurrentRpcs()` is to start a specific number of concurrent RPCs, so changing it to start less than that number is basically breaking it.

If we have tests where we're hitting the max_concurrent_requests limit, then let's instead change those tests to raise that limit in the xDS config.

**Review by markdroth — CHANGES_REQUESTED** [review]

File: test/cpp/end2end/xds_end2end_test.cc (hunk 1 above)
Comment (lidizheng) [reply test/cpp/end2end/xds_end2end_test.cc:2683]: The suggestion is implemented in https://github.com/grpc/grpc/pull/26626.

My original goal of `SendConcurrentRpcs` is sending a certain number of RPCs in a concurrent fashion, so the test cases can run faster. What do you think about adding a `SendConcurrentRpcsWithThrottle` method?

Comment (markdroth) [reply test/cpp/end2end/xds_end2end_test.cc:2683]: I wouldn't be opposed to it if we had a need for it.  But I don't really see the need.

I think #26626 looks good.

**Comment by lidizheng** [comment]
Closing this PR, since https://github.com/grpc/grpc/pull/26626 is a better solution.

**Pull request closed without merging**